### PR TITLE
S8t1

### PR DESCRIPTION
**Pull request by AndyB0o1**

### CODE REVIEW

File: app/Http/Controllers/BookController.php [hunk 1]
```diff
@@ -146,4 +146,40 @@ public function returnBook(int $id, Request $request)
             'success' => true
         ]);
     }
+    public function addBook(Request $request)
+    {
+        $request->validate([
+            'title' => 'required|string',
+            'author' => 'required|string',
```
Comment (ashleycoles) [inline app/Http/Controllers/BookController.php]: Just to be on the safe side, can you get some max length checks on this to line up with your db column lengths

File: app/Http/Controllers/BookController.php [hunk 2]
```diff
@@ -146,4 +146,40 @@ public function returnBook(int $id, Request $request)
...
+        $book->author = $request->author;
+        $book->year = $request->year;
+        $book->blurb = $request->blurb;
+        $book->image = $request->image;
+        $book->page_count = $request->page_count;
+        $book->claimed = 0;
+        $book->genre_id = $request->genre_id;
+
+        if ($book->save())
+        {
```
Comment (ashleycoles) [inline app/Http/Controllers/BookController.php]: { on the same line as the if please, with a space between ) {

File: database/migrations/2024_08_13_153202_books_make_fields_nullable.php [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+            $table->string('image')->nullable()->change();
+        });
+    }
+
+    /**
+     * Reverse the migrations.
+     */
+    public function down(): void
+    {
+        //
```
Comment (ashleycoles) [inline database/migrations/2024_08_13_153202_books_make_fields_nullable.php:27]: Usually I would complain about this - but trying to reverse making them nullable might actually not work. So in this rare case, I think it's ok to leave down empty

File: tests/Feature/BookTest.php [hunk 4]
```diff
@@ -249,4 +249,44 @@ public function test_getBookGenre_success(): void
...
+        $testData = [
+            'title' => 'A boooooook'
+        ];
+
+        $response = $this->postJson('/api/books', $testData);
+
+        $response->assertStatus(422)
+        ->assertJson(function(AssertableJson $json) {
+            $json->hasAll(['message', 'errors']);
+        });
```
Comment (ashleycoles) [inline tests/Feature/BookTest.php]: You should use assertInvalid here instead - it'll automatically check for specific errors

File: tests/Feature/BookTest.php [hunk 5]
```diff
@@ -249,4 +249,44 @@ public function test_getBookGenre_success(): void
                     });
             });
     }
+
+    public function test_addBook_success(): void
```
Comment (ashleycoles) [inline tests/Feature/BookTest.php:253]: You should add another one very similar to this, but with all of the book data

File: tests/Feature/BookTest.php [hunk 6]
```diff
@@ -249,4 +249,44 @@ public function test_getBookGenre_success(): void
...
+
+        $response->assertStatus(201)
+        ->assertJson(function (AssertableJson $json) {
+            $json->hasAll(['message', 'success']);
+        });
+
+        $this->assertDatabaseHas('books', $testData );
+    }
+
+    public function test_addBook_failure(): void
```
Comment (ashleycoles) [inline tests/Feature/BookTest.php]: Can you add an extra test to test the validation a bit more - by providing invalid data for all fields

File: tests/Feature/BookTest.php (hunk 4 above)
Comment (ashleycoles) [reply tests/Feature/BookTest.php]: https://laravel.com/docs/11.x/http-tests#validation-assert-invalid

**Review by ashleycoles — CHANGES_REQUESTED** [review]

File: database/migrations/2024_08_13_153202_books_make_fields_nullable.php (hunk 3 above)
Comment (Alathea95) [reply database/migrations/2024_08_13_153202_books_make_fields_nullable.php:27]: thanks for the feedback ash!

**Review by ashleycoles — APPROVED** [review]

**Pull request merged**